apps/builder/app/builder/shared/expression-editor.tsx

```diff
@@ -80,13 +85,95 @@ const VariablesData = Facet.define<{
 // completion based on
 // https://github.com/codemirror/lang-javascript/blob/4dcee95aee9386fd2c8ad55f93e587b39d968489/src/complete.ts
 
+const Identifier = /^[\p{L}$][\p{L}\p{N}$]*$/u;
+
+const pathFor = (
+  read: (node: SyntaxNode) => string,
+  member: SyntaxNode,
+  name: string
+) => {
+  const path: string[] = [];
+  // traverse from current node to the root variable
+  for (;;) {
+    const object = member.firstChild;
+    if (object?.name === "VariableName") {
+      path.push(read(object));
+      return { path: path.reverse(), name };
+    }
+    if (object?.name === "MemberExpression") {
+      // MemberExpression(SyntaxNode PropertyName)
+      if (object.lastChild?.name === "PropertyName") {
+        path.push(read(object.lastChild!));
+        member = object;
+        continue;
+      }
+      // MemberExpression(SyntaxNode [ SyntaxNode ])
+      if (object.lastChild?.name === "]") {
+        const computed = object.lastChild.prevSibling;
+        if (computed?.name === "Number") {
+          path.push(read(computed));
+          member = object;
+          continue;
+        }
+        if (computed?.name === "String") {
+          // trim quotes from string literal
+          path.push(read(computed).slice(1, -1));
+          member = object;
+          continue;
+        }
+      }
+    }
+    // unexpected case
+    break;
+  }
+};
+
+/// Helper function for defining JavaScript completion sources. It
+/// returns the completable name and object path for a completion
+/// context, or undefined if no name/property completion should happen at
+/// that position. For example, when completing after `a.b.c` it will
+/// return `{path: ["a", "b"], name: "c"}`. When completing after `x`
+/// it will return `{path: [], name: "x"}`. When not in a property or
+/// name, it will return undefined if `context.explicit` is false, and
+/// `{path: [], name: ""}` otherwise.
+const completionPath = (
```

Comment:
lets add some tests for this?

Reply:
Have no idea how to test this 😄

Reply:
isn't there a way to get that object from codemirror? check how they test it? Maybe its not too hard to just take a snapshot of it from from the browser?

Reply:
I would prefer to spend time on porting it to original package